include/picongpu/simulationControl/AdjustDomainSize.hpp

```diff
@@ -0,0 +1,363 @@
+            }
+
+            /* since we are not doing independent reduces per slice we need
+             * to adjust the offset result by dividing with the number of
+             * MPI ranks in all other dimensions.
+             */
+            offset /= static_cast< uint64_t >( m_numDevices.productOfComponents() / m_numDevices[ dim ] );
+            m_localDomainOffset[ dim ] = static_cast< int >( offset );
+
+        }
+
+        /** ensure that the local size is a multiple of the supercell size
+         *
+         * @param dim dimension to update
+         */
+        void multipleOfSuperCell( size_t const dim )
+        {
+            int const sCellSize = SuperCellSize::toRT()[ dim ];
+            // round up to full supercells
+            int const validLocalSize =
+                ( ( m_localDomainSize[ dim ] + sCellSize - 1 ) / sCellSize ) *
+                sCellSize;
+
+            if( validLocalSize != m_localDomainSize[ dim ] )
+            {
+                std::cout << "Dimension " << dim <<
+                    ": local grid size is not a multiple of supercell size. Adjust from " <<
+                    m_localDomainSize[ dim ] << " to " <<  validLocalSize << std::endl;
+                sizeChanged = true;
+            }
+
+            m_localDomainSize[ dim ] = validLocalSize;
+        }
+
+        /** ensure that we have a CORE and BORDER region
+         *
+         * Each region must have the size of at least one supercell.
+         *
+         * @param dim dimension to update
+         */
+        void minThreeSuperCells( size_t dim )
+        {
+            int superCellCount = m_localDomainSize[ dim ] / SuperCellSize::toRT()[ dim ];
+            if( superCellCount < 3 )
+                superCellCount = 3;
+
+            int validLocalSize = superCellCount * SuperCellSize::toRT()[ dim ];
+            if( validLocalSize != m_localDomainSize[ dim ] )
+            {
+                std::cout << "Dimension " << dim <<
+                    ": local grid size is not contains at least 3 supercells. Adjust from " <<
+                    m_localDomainSize[ dim ] << " to " <<  validLocalSize << std::endl;
+                sizeChanged = true;
+            }
+
+            m_localDomainSize[ dim ] = validLocalSize;
+        }
+
+        /** ensure that the absorber fits into the local domain
+         *
+         * The methods checks the local domain size only if the absorber for the
+         * given dimension is enabled.
+         *
+         * @param dim dimension to update
+         */
+        void greaterEqualThanAbsorber( size_t const dim )
+        {
+            int validLocalSize = m_localDomainSize[ dim ];
+            if(
+                !m_isPeriodic[ dim ] &&
+                // device has absorber
+                ( m_mpiPosition[ dim ] == 0 || m_mpiPosition[ dim ] == m_numDevices[ dim ] - 1 )
+            )
+            {
+                int const maxAbsorberCells = static_cast< int >(
```

Comment:
I might have misunderstood it, but the logic of taking the maximum seems wrong to me. What if we have multiple devices along the current direction and the right-side absorber is longer (more cells) than left-side. Then it is possible that the left-most local domain is large enough to fit the left absorber, but not enough for the right absorber, and the right-most local domain is large enough. As I understand it, this case is perfectly fine, but the check would fail.

Reply:
That is correct I will update this part and handle the cases correct.